File: rust/src/session.rs
// Copyright (c) Microsoft Corporation. All rights reserved.

//! CopilotSession - represents a single conversation session with the Copilot CLI.
//!
//! Sessions are created via [`CopilotClient::create_session`] or resumed via
//! [`CopilotClient::resume_session`]. They maintain conversation state, handle events,
//! and manage tool execution.

use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

use serde_json::Value;
use tokio::sync::{mpsc, Mutex};

use crate::jsonrpc::JsonRpcClient;
use crate::types::*;
use crate::CopilotError;

// ============================================================================
// Handler Type Aliases
// ============================================================================

/// A tool handler is an async function that takes arguments and invocation info,
/// and returns a result value.
pub type ToolHandler = Arc<
    dyn Fn(
            Value,
            ToolInvocation,
        ) -> Pin<Box<dyn Future<Output = Result<Value, CopilotError>> + Send>>
        + Send
        + Sync,
>;

/// A permission handler is an async function that takes a permission request
/// and returns a permission result.
pub type PermissionHandlerFn = Arc<
    dyn Fn(
            PermissionRequest,
            String,
        ) -> Pin<Box<dyn Future<Output = Result<PermissionRequestResult, CopilotError>> + Send>>
        + Send
        + Sync,
>;

/// A user input handler is an async function that takes a user input request
/// and returns a response.
pub type UserInputHandlerFn = Arc<
    dyn Fn(
            UserInputRequest,
            String,
        ) -> Pin<Box<dyn Future<Output = Result<UserInputResponse, CopilotError>> + Send>>
        + Send
        + Sync,
>;

/// A hooks handler is an async function that takes hook type, input, and session ID,
/// and returns optional output.
pub type HooksHandlerFn = Arc<
    dyn Fn(
            String,
            Value,
            String,
        ) -> Pin<Box<dyn Future<Output = Result<Option<Value>, CopilotError>> + Send>>
        + Send
        + Sync,
>;

/// A session event handler callback.
pub type SessionEventHandlerFn = Arc<dyn Fn(SessionEvent) + Send + Sync>;

/// A typed session event handler for a specific event type.
pub type TypedSessionEventHandlerFn = Arc<dyn Fn(SessionEvent) + Send + Sync>;

// ============================================================================
// Unsubscribe Guard
// ============================================================================

/// An RAII guard that unsubscribes a handler when dropped.
/// Can also be explicitly unsubscribed via `unsubscribe()`.
pub struct Subscription {
    unsubscribe_fn: Option<Box<dyn FnOnce() + Send>>,
}

impl Subscription {
    fn new(f: impl FnOnce() + Send + 'static) -> Self {
        Self {
            unsubscribe_fn: Some(Box::new(f)),
        }
    }

    /// Explicitly unsubscribes this handler.
    pub fn unsubscribe(mut self) {
        if let Some(f) = self.unsubscribe_fn.take() {
            f();
        }
    }
}

impl Drop for Subscription {
    fn drop(&mut self) {
        if let Some(f) = self.unsubscribe_fn.take() {
            f();
        }
    }
}

// ============================================================================
// CopilotSession
// ============================================================================

/// Represents a single conversation session with the Copilot CLI.
///
/// A session maintains conversation state, handles events, and manages tool execution.
///
/// # Examples
///
/// ```rust,no_run
/// # use copilot_sdk::*;
/// # async fn example() -> Result<(), CopilotError> {
/// let client = CopilotClient::new(CopilotClientOptions::default());
/// let session = client.create_session(SessionConfig::default()).await?;
///
/// // Subscribe to events
/// let sub = session.on(|event| {
///     if event.is_assistant_message() {
///         if let Some(content) = event.assistant_message_content() {
///             println!("Assistant: {}", content);
///         }
///     }
/// }).await;
///
/// // Send a message and wait for completion
/// let response = session.send_and_wait(
///     MessageOptions { prompt: "Hello!".into(), attachments: None, mode: None },
///     None,
/// ).await?;
///
/// // Clean up
/// session.destroy().await?;
/// # Ok(())
/// # }
/// ```
pub struct CopilotSession {
    /// The unique session ID.
    session_id: String,
    /// Path to the session workspace directory (when infinite sessions are enabled).
    workspace_path: Option<String>,
    /// Reference to the JSON-RPC client.
    rpc_client: Arc<JsonRpcClient>,
    /// Registered tool handlers, keyed by tool name.
    tool_handlers: Arc<Mutex<HashMap<String, ToolHandler>>>,
    /// Permission request handler.
    permission_handler: Arc<Mutex<Option<PermissionHandlerFn>>>,
    /// User input request handler.
    user_input_handler: Arc<Mutex<Option<UserInputHandlerFn>>>,
    /// Hooks handler.
    hooks_handler: Arc<Mutex<Option<HooksHandlerFn>>>,
    /// Wildcard event handlers.
    event_handlers: Arc<Mutex<Vec<(u64, SessionEventHandlerFn)>>>,
    /// Typed event handlers, keyed by event type string.
    typed_event_handlers: Arc<Mutex<HashMap<String, Vec<(u64, TypedSessionEventHandlerFn)>>>>,
    /// Counter for handler IDs (for unsubscribe).
    next_handler_id: Arc<Mutex<u64>>,
}

impl CopilotSession {
    /// Creates a new CopilotSession. This is called internally by CopilotClient.
    pub(crate) fn new(
        session_id: String,
        rpc_client: Arc<JsonRpcClient>,
        workspace_path: Option<String>,
    ) -> Self {
        Self {
            session_id,
            workspace_path,
            rpc_client,
            tool_handlers: Arc::new(Mutex::new(HashMap::new())),
            permission_handler: Arc::new(Mutex::new(None)),
            user_input_handler: Arc::new(Mutex::new(None)),
            hooks_handler: Arc::new(Mutex::new(None)),
            event_handlers: Arc::new(Mutex::new(Vec::new())),
            typed_event_handlers: Arc::new(Mutex::new(HashMap::new())),
            next_handler_id: Arc::new(Mutex::new(0)),
        }
    }

    /// Returns the session ID.
    pub fn session_id(&self) -> &str {
        &self.session_id
    }

    /// Returns the workspace path (when infinite sessions are enabled).
    pub fn workspace_path(&self) -> Option<&str> {
        self.workspace_path.as_deref()
    }

    // ========================================================================
    // Message Sending
    // ========================================================================

    /// Sends a message to this session.
    ///
    /// The message is processed asynchronously. Subscribe to events via `on()`
    /// to receive streaming responses and other session events.
    ///
    /// Returns the message ID.
    pub async fn send(&self, options: MessageOptions) -> Result<String, CopilotError> {
        let params = serde_json::json!({
            "sessionId": self.session_id,
            "prompt": options.prompt,
            "attachments": options.attachments,
            "mode": options.mode,
        });

        let response = self.rpc_client.request("session.send", params, None).await?;
        let message_id = response
            .get("messageId")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        Ok(message_id)
    }

    /// Sends a message and waits until the session becomes idle.
    ///
    /// This combines `send()` with waiting for the `session.idle` event.
    /// Returns the last `assistant.message` event received, or `None`.
    ///
    /// # Arguments
    /// * `options` - The message options
    /// * `timeout` - Optional timeout in milliseconds (defaults to 60000)
    pub async fn send_and_wait(
        &self,
        options: MessageOptions,
        timeout: Option<u64>,
    ) -> Result<Option<SessionEvent>, CopilotError> {
        let effective_timeout = timeout.unwrap_or(60_000);

        // Channel to signal idle or error
        let (idle_tx, mut idle_rx) = mpsc::channel::<Result<(), CopilotError>>(1);
        let last_assistant_message: Arc<Mutex<Option<SessionEvent>>> =
            Arc::new(Mutex::new(None));

        let last_msg_clone = Arc::clone(&last_assistant_message);
        let idle_tx_clone = idle_tx.clone();

        // Register event handler BEFORE calling send to avoid race condition
        let sub = self
            .on(move |event: SessionEvent| {
                if event.is_assistant_message() {
                    let mut msg = last_msg_clone.blocking_lock();
                    *msg = Some(event);
                } else if event.is_session_idle() {
                    let _ = idle_tx_clone.try_send(Ok(()));
                } else if event.is_session_error() {
                    let error_msg = event
                        .error_message()
                        .unwrap_or("Unknown error")
                        .to_string();
                    let _ = idle_tx_clone.try_send(Err(CopilotError::SessionError(error_msg)));
                }
            })
            .await;

        // Send the message
        self.send(options).await?;

        // Wait for idle or timeout
        let result = tokio::time::timeout(
            std::time::Duration::from_millis(effective_timeout),
            idle_rx.recv(),
        )
        .await;

        // Unsubscribe
        sub.unsubscribe();

        match result {
            Ok(Some(Ok(()))) => {
                let msg = last_assistant_message.lock().await;
                Ok(msg.clone())
            }
            Ok(Some(Err(e))) => Err(e),
            Ok(None) => Err(CopilotError::ConnectionClosed),
            Err(_) => Err(CopilotError::Timeout(effective_timeout)),
        }
    }

    // ========================================================================
    // Event Subscription
    // ========================================================================

    /// Subscribes to all events from this session.
    ///
    /// Returns a `Subscription` that unsubscribes when dropped or when
    /// `unsubscribe()` is called.
    pub async fn on<F>(&self, handler: F) -> Subscription
    where
        F: Fn(SessionEvent) + Send + Sync + 'static,
    {
        let handler_id = {
            let mut id = self.next_handler_id.lock().await;
            *id += 1;
            *id
        };

        let handler_arc: SessionEventHandlerFn = Arc::new(handler);
        {
            let mut handlers = self.event_handlers.lock().await;
            handlers.push((handler_id, handler_arc));
        }

        let event_handlers = Arc::clone(&self.event_handlers);
        Subscription::new(move || {
            // We need to use blocking_lock since unsubscribe may be called from Drop
            // in a non-async context
            let mut handlers = event_handlers.blocking_lock();
            handlers.retain(|(id, _)| *id != handler_id);
        })
    }

    /// Subscribes to a specific event type from this session.
    ///
    /// # Arguments
    /// * `event_type` - The event type string (e.g., "assistant.message", "session.idle")
    /// * `handler` - The callback function
    pub async fn on_event<F>(&self, event_type: &str, handler: F) -> Subscription
    where
        F: Fn(SessionEvent) + Send + Sync + 'static,
    {
        let handler_id = {
            let mut id = self.next_handler_id.lock().await;
            *id += 1;
            *id
        };

        let handler_arc: TypedSessionEventHandlerFn = Arc::new(handler);
        let event_type_str = event_type.to_string();
        {
            let mut handlers = self.typed_event_handlers.lock().await;
            handlers
                .entry(event_type_str.clone())
                .or_default()
                .push((handler_id, handler_arc));
        }

        let typed_handlers = Arc::clone(&self.typed_event_handlers);
        let et = event_type_str;
        Subscription::new(move || {
            let mut handlers = typed_handlers.blocking_lock();
            if let Some(list) = handlers.get_mut(&et) {
                list.retain(|(id, _)| *id != handler_id);
            }
        })
    }

    // ========================================================================
    // Event Dispatch (internal, called by CopilotClient)
    // ========================================================================

    /// Dispatches a session event to all registered handlers.
    pub(crate) async fn dispatch_event(&self, event: SessionEvent) {
        // Dispatch to typed handlers
        {
            let handlers = self.typed_event_handlers.lock().await;
            if let Some(list) = handlers.get(&event.event_type) {
                for (_, handler) in list {
                    handler(event.clone());
                }
            }
        }

        // Dispatch to wildcard handlers
        {
            let handlers = self.event_handlers.lock().await;
            for (_, handler) in handlers.iter() {
                handler(event.clone());
            }
        }
    }

    // ========================================================================
    // Tool Registration
    // ========================================================================

    /// Registers a tool handler.
    pub async fn register_tool(&self, name: &str, handler: ToolHandler) {
        let mut handlers = self.tool_handlers.lock().await;
        handlers.insert(name.to_string(), handler);
    }

    /// Registers multiple tool handlers.
    pub async fn register_tools(&self, tools: Vec<(String, ToolHandler)>) {
        let mut handlers = self.tool_handlers.lock().await;
        handlers.clear();
        for (name, handler) in tools {
            handlers.insert(name, handler);
        }
    }

    /// Gets a registered tool handler by name.
    pub(crate) async fn get_tool_handler(&self, name: &str) -> Option<ToolHandler> {
        let handlers = self.tool_handlers.lock().await;
        handlers.get(name).cloned()
    }

    // ========================================================================
    // Permission Handler
    // ========================================================================

    /// Registers a permission request handler.
    pub async fn register_permission_handler(&self, handler: PermissionHandlerFn) {
        let mut h = self.permission_handler.lock().await;
        *h = Some(handler);
    }

    /// Handles an incoming permission request from the server.
    pub(crate) async fn handle_permission_request(
        &self,
        request: Value,
    ) -> Result<PermissionRequestResult, CopilotError> {
        let handler = self.permission_handler.lock().await;
        if let Some(ref h) = *handler {
            let perm_request: PermissionRequest = serde_json::from_value(request)
                .map_err(|e| CopilotError::Serialization(e.to_string()))?;
            h(perm_request, self.session_id.clone()).await
        } else {
            Ok(PermissionRequestResult {
                kind: PermissionResultKind::DeniedNoApprovalRuleAndCouldNotRequestFromUser,
                rules: None,
            })
        }
    }

    // ========================================================================
    // User Input Handler
    // ========================================================================

    /// Registers a user input request handler.
    pub async fn register_user_input_handler(&self, handler: UserInputHandlerFn) {
        let mut h = self.user_input_handler.lock().await;
        *h = Some(handler);
    }

    /// Handles an incoming user input request from the server.
    pub(crate) async fn handle_user_input_request(
        &self,
        request: Value,
    ) -> Result<UserInputResponse, CopilotError> {
        let handler = self.user_input_handler.lock().await;
        if let Some(ref h) = *handler {
            let input_request: UserInputRequest = serde_json::from_value(request)
                .map_err(|e| CopilotError::Serialization(e.to_string()))?;
            h(input_request, self.session_id.clone()).await
        } else {
            Err(CopilotError::NoHandler(
                "User input requested but no handler registered".to_string(),
            ))
        }
    }

    // ========================================================================
    // Hooks Handler
    // ========================================================================

    /// Registers a hooks handler for all hook types.
    pub async fn register_hooks_handler(&self, handler: HooksHandlerFn) {
        let mut h = self.hooks_handler.lock().await;
        *h = Some(handler);
    }

    /// Handles an incoming hooks invocation from the server.
    pub(crate) async fn handle_hooks_invoke(
        &self,
        hook_type: &str,
        input: Value,
    ) -> Result<Option<Value>, CopilotError> {
        let handler = self.hooks_handler.lock().await;
        if let Some(ref h) = *handler {
            h(hook_type.to_string(), input, self.session_id.clone()).await
        } else {
            Ok(None)
        }
    }

    // ========================================================================
    // Session Operations
    // ========================================================================

    /// Retrieves all events and messages from this session's history.
    pub async fn get_messages(&self) -> Result<Vec<SessionEvent>, CopilotError> {
        let params = serde_json::json!({ "sessionId": self.session_id });
        let response = self
            .rpc_client
            .request("session.getMessages", params, None)
            .await?;
        let events: Vec<SessionEvent> = serde_json::from_value(
            response
                .get("events")
                .cloned()
                .unwrap_or(Value::Array(vec![])),
        )
        .map_err(|e| CopilotError::Serialization(e.to_string()))?;
        Ok(events)
    }

    /// Destroys this session and releases all associated resources.
    ///
    /// After calling this method, the session can no longer be used.
    pub async fn destroy(&self) -> Result<(), CopilotError> {
        let params = serde_json::json!({ "sessionId": self.session_id });
        self.rpc_client
            .request("session.destroy", params, None)
            .await?;

        // Clear all handlers
        {
            let mut handlers = self.event_handlers.lock().await;
            handlers.clear();
        }
        {
            let mut handlers = self.typed_event_handlers.lock().await;
            handlers.clear();
        }
        {
            let mut handlers = self.tool_handlers.lock().await;
            handlers.clear();
        }
        {
            let mut handler = self.permission_handler.lock().await;
            *handler = None;
        }
        {
            let mut handler = self.user_input_handler.lock().await;
            *handler = None;
        }

        Ok(())
    }

    /// Aborts the currently processing message in this session.
    pub async fn abort(&self) -> Result<(), CopilotError> {
        let params = serde_json::json!({ "sessionId": self.session_id });
        self.rpc_client
            .request("session.abort", params, None)
            .await?;
        Ok(())
    }
}

File: r/R/session.R
#' CopilotSession
#'
#' Represents a single conversation session with the Copilot CLI.
#'
#' @description
#' A session maintains conversation state, handles events, and manages tool execution.
#' Sessions are created via \code{CopilotClient$create_session()} or resumed via
#' \code{CopilotClient$resume_session()}.
#'
#' @export
CopilotSession <- R6::R6Class(
  "CopilotSession",
  public = list(
    #' @field session_id Character. The unique identifier for this session.
    session_id = NULL,

    #' @field workspace_path Character or NULL. Path to workspace directory
    #'   when infinite sessions are enabled.
    workspace_path = NULL,

    #' @description Create a new CopilotSession.
    #'
    #' Note: This constructor is internal. Use CopilotClient$create_session()
    #' to create sessions.
    #'
    #' @param session_id Character. Session identifier.
    #' @param client JsonRpcClient R6 object.
    #' @param workspace_path Character or NULL.
    initialize = function(session_id, client, workspace_path = NULL) {
      self$session_id <- session_id
      self$workspace_path <- workspace_path
      private$client <- client
      private$event_handlers <- list()
      private$tool_handlers <- list()
      private$permission_handler <- NULL
      private$user_input_handler <- NULL
      private$hooks <- NULL
    },

    #' @description Send a message to this session.
    #'
    #' @param prompt Character. The prompt/message text to send.
    #' @param attachments List or NULL. Optional file/directory attachments.
    #' @param mode Character or NULL. "enqueue" or "immediate".
    #' @param response_format Character or NULL. Response format ("text", "image", or "json_object").
    #' @param image_options List or NULL. Image generation options (from \code{image_options()}).
    #'
    #' @return Character. The message ID of the response.
    send = function(prompt, attachments = NULL, mode = NULL,
                    response_format = NULL, image_options = NULL) {
      payload <- list(
        sessionId = self$session_id,
        prompt = prompt
      )
      if (!is.null(attachments)) payload$attachments <- attachments
      if (!is.null(mode)) payload$mode <- mode
      if (!is.null(response_format)) payload$responseFormat <- response_format
      if (!is.null(image_options)) payload$imageOptions <- image_options

      response <- private$client$request("session.send", payload)
      response$messageId
    },

    #' @description Send a message and wait until the session becomes idle.
    #'
    #' This is a convenience method that sends a message and then polls for
    #' events until session.idle is received. The event handlers registered
    #' via \code{on()} are still called for each event during the wait.
    #'
    #' @param prompt Character. The prompt/message text.
    #' @param attachments List or NULL. Optional attachments.
    #' @param mode Character or NULL. Processing mode.
    #' @param response_format Character or NULL. Response format ("text", "image", or "json_object").
    #' @param image_options List or NULL. Image generation options (from \code{image_options()}).
    #' @param timeout Numeric. Timeout in seconds (default 60).
    #'
    #' @return The last assistant.message SessionEvent, or NULL if none received.
    send_and_wait = function(prompt, attachments = NULL, mode = NULL,
                             response_format = NULL, image_options = NULL,
                             timeout = 60) {
      last_assistant_message <- NULL
      error_event <- NULL
      idle_received <- FALSE

      # Temporary event handler to track idle and messages
      temp_handler <- function(event) {
        if (identical(event$type, "assistant.message")) {
          last_assistant_message <<- event
        } else if (identical(event$type, "session.idle")) {
          idle_received <<- TRUE
        } else if (identical(event$type, "session.error")) {
          error_msg <- if (!is.null(event$data$message)) event$data$message else "Unknown error"
          error_event <<- paste0("Session error: ", error_msg)
          idle_received <<- TRUE
        }
      }

      unsubscribe <- self$on(temp_handler)
      on.exit(unsubscribe(), add = TRUE)

      # Send the message
      self$send(prompt = prompt, attachments = attachments, mode = mode,
                response_format = response_format, image_options = image_options)

      # Poll until idle or timeout
      start_time <- proc.time()[["elapsed"]]
      while (!idle_received) {
        elapsed <- proc.time()[["elapsed"]] - start_time
        if (elapsed > timeout) {
          stop(paste0("Timeout after ", timeout, "s waiting for session.idle"))
        }
        # Poll processes incoming messages (notifications, requests)
        private$client$poll(timeout_ms = 500L)
      }

      if (!is.null(error_event)) {
        stop(error_event)
      }

      last_assistant_message
    },

    #' @description Subscribe to events from this session.
    #'
    #' @param handler Function(event). Called with each SessionEvent.
    #'
    #' @return A function that when called, unsubscribes the handler.
    on = function(handler) {
      private$event_handlers <- c(private$event_handlers, list(handler))
      local_handler <- handler
      function() {
        private$event_handlers <- Filter(
          function(h) !identical(h, local_handler),
          private$event_handlers
        )
      }
    },

    #' @description Dispatch an event to all registered handlers.
    #' @param event SessionEvent R6 object.
    #' @keywords internal
    dispatch_event = function(event) {
      # Handle elicitation.requested events internally
      if (identical(event$type, "elicitation.requested") &&
          !is.null(private$elicitation_handler)) {
        private$handle_elicitation_request(event$data)
      }

      for (handler in private$event_handlers) {
        tryCatch(
          handler(event),
          error = function(e) {
            message(paste("Error in session event handler:", e$message))
          }
        )
      }
    },

    #' @description Register custom tool handlers.
    #' @param tools List of Tool R6 objects or NULL.
    #' @keywords internal
    register_tools = function(tools) {
      private$tool_handlers <- list()
      if (is.null(tools)) return(invisible(NULL))
      for (tool in tools) {
        if (!is.null(tool$name) && !is.null(tool$handler)) {
          private$tool_handlers[[tool$name]] <- tool$handler
        }
      }
    },

    #' @description Get a registered tool handler by name.
    #' @param name Character. Tool name.
    #' @return Function or NULL.
    #' @keywords internal
    get_tool_handler = function(name) {
      private$tool_handlers[[name]]
    },

    #' @description Register a permission request handler.
    #' @param handler Function or NULL.
    #' @keywords internal
    register_permission_handler = function(handler) {
      private$permission_handler <- handler
    },

    #' @description Handle a permission request from the CLI.
    #' @param request Named list. The permission request data.
    #' @return Named list with permission decision.
    #' @keywords internal
    handle_permission_request = function(request) {
      if (is.null(private$permission_handler)) {
        return(list(kind = "denied-no-approval-rule-and-could-not-request-from-user"))
      }

      tryCatch(
        {
          result <- private$permission_handler(
            request,
            list(session_id = self$session_id)
          )
          if (inherits(result, "PermissionRequestResult")) {
            result$to_list()
          } else {
            result
          }
        },
        error = function(e) {
          list(kind = "denied-no-approval-rule-and-could-not-request-from-user")
        }
      )
    },

    #' @description Register a user input handler.
    #' @param handler Function or NULL.
    #' @keywords internal
    register_user_input_handler = function(handler) {
      private$user_input_handler <- handler
    },

    #' @description Handle a user input request from the CLI.
    #' @param request Named list. The user input request data.
    #' @return Named list with answer and was_freeform.
    #' @keywords internal
    handle_user_input_request = function(request) {
      if (is.null(private$user_input_handler)) {
        stop("User input requested but no handler registered")
      }

      input_req <- UserInputRequest$new(
        question = request$question %||% "",
        choices = request$choices,
        allow_freeform = if (!is.null(request$allowFreeform)) request$allowFreeform else TRUE
      )

      result <- private$user_input_handler(
        input_req,
        list(session_id = self$session_id)
      )

      if (inherits(result, "UserInputResponse")) {
        list(answer = result$answer, was_freeform = result$was_freeform)
      } else {
        list(answer = result$answer, was_freeform = isTRUE(result$wasFreeform %||% result$was_freeform))
      }
    },

    #' @description Register hook handlers.
    #' @param hooks Named list of hook handler functions.
    #' @keywords internal
    register_hooks = function(hooks) {
      private$hooks <- hooks
    },

    #' @description Register an elicitation request handler.
    #' @param handler Function or NULL. Called with (context, request_id).
    #' @keywords internal
    register_elicitation_handler = function(handler) {
      private$elicitation_handler <- handler
    },

    #' @description Register slash commands for this session.
    #' @param commands List of CommandDefinition objects or NULL.
    #' @keywords internal
    register_commands = function(commands) {
      private$commands <- commands
    },

    #' @description Handle a hooks invocation from the CLI.
    #' @param hook_type Character. Hook type.
    #' @param input_data Any. Hook input.
    #' @return Hook output or NULL.
    #' @keywords internal
    handle_hooks_invoke = function(hook_type, input_data) {
      if (is.null(private$hooks)) return(NULL)

      handler_map <- list(
        preToolUse = private$hooks$on_pre_tool_use,
        postToolUse = private$hooks$on_post_tool_use,
        userPromptSubmitted = private$hooks$on_user_prompt_submitted,
        sessionStart = private$hooks$on_session_start,
        sessionEnd = private$hooks$on_session_end,
        errorOccurred = private$hooks$on_error_occurred
      )

      handler <- handler_map[[hook_type]]
      if (is.null(handler)) return(NULL)

      tryCatch(
        handler(input_data, list(session_id = self$session_id)),
        error = function(e) NULL
      )
    },

    #' @description Retrieve metadata for this session.
    #' @return A named list containing the session metadata.
    get_metadata = function() {
      private$client$request(
        "session.getMetadata",
        list(sessionId = self$session_id)
      )
    },

    #' @description Retrieve all events/messages from this session's history.
    #' @return A list of SessionEvent R6 objects.
    get_messages = function() {
      response <- private$client$request(
        "session.getMessages",
        list(sessionId = self$session_id)
      )
      events_dicts <- response$events %||% list()
      lapply(events_dicts, session_event_from_list)
    },

    #' @description Destroy this session and release resources.
    destroy = function() {
      private$client$request("session.destroy", list(sessionId = self$session_id))
      private$event_handlers <- list()
      private$tool_handlers <- list()
      private$permission_handler <- NULL
      private$user_input_handler <- NULL
      private$hooks <- NULL
      invisible(NULL)
    },

    #' @description Abort the currently processing message.
    abort = function() {
      private$client$request("session.abort", list(sessionId = self$session_id))
      invisible(NULL)
    }
  ),

  private = list(
    client = NULL,
    event_handlers = NULL,
    tool_handlers = NULL,
    permission_handler = NULL,
    user_input_handler = NULL,
    hooks = NULL,
    elicitation_handler = NULL,
    commands = NULL,

    handle_elicitation_request = function(data) {
      request_id <- data$requestId
      context <- list(
        session_id = self$session_id,
        message = data$message,
        requested_schema = data$requestedSchema,
        mode = data$mode,
        elicitation_source = data$elicitationSource,
        url = data$url
      )

      tryCatch(
        {
          result <- private$elicitation_handler(context)
          private$client$request("ui.handlePendingElicitation", list(
            requestId = request_id,
            result = result
          ))
        },
        error = function(e) {
          tryCatch(
            private$client$request("ui.handlePendingElicitation", list(
              requestId = request_id,
              result = list(action = "cancel")
            )),
            error = function(e2) NULL
          )
        }
      )
    }
  )
)
